File: src/main/java/mediathek/tool/subtitles/ttml2/SubRipHtmlExporter.java
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.tool.subtitles.ttml2;

import mediathek.tool.subtitles.ttml2.SubtitleDocument.Cue;
import mediathek.tool.subtitles.ttml2.SubtitleDocument.StyledRun;
import mediathek.tool.subtitles.ttml2.SubtitleDocument.TextStyle;

import java.time.Duration;

/**
 * SubRip exporter with HTML-like style tags:
 * - <b>, <i>, <u>
 * - <font color="#RRGGBB"> (alpha ignored; SRT has no standard alpha support)
 */
public final class SubRipHtmlExporter {

    private static String openTags(TextStyle s) {
        if (s == null)
            return "";
        StringBuilder sb = new StringBuilder();
        if (s.bold())
            sb.append("<b>");
        if (s.italic())
            sb.append("<i>");
        if (s.underline())
            sb.append("<u>");
        if (s.color() != null && !s.color().isTransparent()) {
            sb.append("<font color=\"").append(String.format("#%02X%02X%02X", s.color().r(), s.color().g(), s.color().b())).append("\">");
        }
        return sb.toString();
    }

    private static String closeTags(TextStyle s) {
        if (s == null)
            return "";
        StringBuilder sb = new StringBuilder();
        if (s.color() != null && !s.color().isTransparent())
            sb.append("</font>");
        if (s.underline())
            sb.append("</u>");
        if (s.italic())
            sb.append("</i>");
        if (s.bold())
            sb.append("</b>");
        return sb.toString();
    }

    private static String escapeHtml(String s) {
        return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;");
    }

    private static String fmt(Duration d) {
        long ms = Math.max(0, d.toMillis());
        long h = ms / 3_600_000;
        ms %= 3_600_000;
        long m = ms / 60_000;
        ms %= 60_000;
        long s = ms / 1_000;
        long r = ms % 1_000;
        return String.format("%02d:%02d:%02d,%03d", h, m, s, r);
    }

    public String export(SubtitleDocument doc) {
        StringBuilder sb = new StringBuilder();
        int idx = 1;

        for (Cue cue : doc.cues()) {
            sb.append(idx++).append('\n');
            sb.append(fmt(cue.start())).append(" --> ").append(fmt(cue.end())).append('\n');

            TextStyle prev = null;
            for (StyledRun run : cue.runs()) {
                TextStyle cur = run.style();

                if (prev != null && !prev.equals(cur))
                    sb.append(closeTags(prev));
                if (prev == null || !prev.equals(cur))
                    sb.append(openTags(cur));

                sb.append(escapeHtml(run.text()).replace("\\n", "\n"));
                prev = cur;
            }
            if (prev != null)
                sb.append(closeTags(prev));

            sb.append("\n\n");
        }
        return sb.toString();
    }
}

File: src/main/kotlin/mediathek/tool/subtitles/ttml2/SubRipHtmlExporter.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.tool.subtitles.ttml2

import mediathek.tool.subtitles.SubtitleDocument
import mediathek.tool.subtitles.SubtitleDocument.TextStyle
import java.time.Duration

/**
 * SubRip exporter with HTML-like style tags:
 * - <b>, <i>, <u>
 * - <font color="#RRGGBB"> (alpha ignored; SRT has no standard alpha support)
 */
class SubRipHtmlExporter {
    fun export(doc: SubtitleDocument): String = buildString {
        var index = 1

        for (cue in doc.cues) {
            append(index++)
            append(LINE_SEPARATOR)
            append(formatTimestamp(cue.start))
            append(" --> ")
            append(formatTimestamp(cue.end))
            append(LINE_SEPARATOR)

            var previousStyle: TextStyle? = null
            for (run in cue.runs) {
                val currentStyle = run.style

                if (previousStyle != null && previousStyle != currentStyle) {
                    append(closeTags(previousStyle))
                }
                if (previousStyle == null || previousStyle != currentStyle) {
                    append(openTags(currentStyle))
                }

                append(escapeText(run.text))
                previousStyle = currentStyle
            }
            if (previousStyle != null) {
                append(closeTags(previousStyle))
            }

            append(LINE_SEPARATOR)
            append(LINE_SEPARATOR)
        }
    }

    private fun openTags(style: TextStyle?): String {
        if (style == null) {
            return ""
        }
        return buildString {
            if (style.bold) {
                append("<b>")
            }
            if (style.italic) {
                append("<i>")
            }
            if (style.underline) {
                append("<u>")
            }
            val color = style.color
            if (color != null && !color.isTransparent()) {
                append("<font color=\"")
                append("#%02X%02X%02X".format(color.r, color.g, color.b))
                append("\">")
            }
        }
    }

    private fun closeTags(style: TextStyle?): String {
        if (style == null) {
            return ""
        }
        return buildString {
            if (style.color != null && !style.color.isTransparent()) {
                append("</font>")
            }
            if (style.underline) {
                append("</u>")
            }
            if (style.italic) {
                append("</i>")
            }
            if (style.bold) {
                append("</b>")
            }
        }
    }

    private fun formatTimestamp(duration: Duration): String {
        var milliseconds = duration.toMillis().coerceAtLeast(0)
        val hours = milliseconds / 3_600_000
        milliseconds %= 3_600_000
        val minutes = milliseconds / 60_000
        milliseconds %= 60_000
        val seconds = milliseconds / 1_000
        val remainingMilliseconds = milliseconds % 1_000
        return "%02d:%02d:%02d,%03d".format(hours, minutes, seconds, remainingMilliseconds)
    }

    private fun escapeText(text: String): String =
        text
            .replace("\\n", LINE_SEPARATOR)
            .replace("<", "&lt;")
            .replace(">", "&gt;")

    private companion object {
        private const val LINE_SEPARATOR = "\r\n"
    }
}
